core/domain/exp_domain.py

```diff
@@ -2964,6 +2967,67 @@ def delete_state(self, state_name):
 
         del self.states[state_name]
 
+
+    def get_trainable_states_dict(self, old_states, change_list):
+        """Retrieves the state names of all trainable states in an exploration
+        segregated into state names with changed and unchanged answer groups.
+
+        Args:
+            old_states: dict. Dictionary containing all State domain objects.
+            change_list: list(dict). A list of changes introduced in this
+                commit.
+
+        Returns:
+            dict. The trainable states dict. This dict has three keys
+                representing state names with changed answer groups, unchanged
+                answer groups and state_name-old_state_name mapping
+                respectively.
+        """
+        trainable_states_dict = {
+            'state_names_with_changed_answer_groups': [],
+            'state_names_with_unchanged_answer_groups': []
+        }
+        new_states = self.states
+
+        renamed_cmds = []
+        for change_dict in change_list:
+            if change_dict['cmd'] == CMD_RENAME_STATE:
+                renamed_cmds.append(change_dict)
+
+        state_names_mapping = {}
+        for state_name in new_states:
+            new_state = new_states[state_name]
+            if not new_state.can_undergo_classification():
+                continue
+
+            old_state_name = state_name
+            for change_dict in reversed(renamed_cmds):
+                if change_dict['new_state_name'] == old_state_name:
+                    old_state_name = change_dict['old_state_name']
+            state_names_mapping[state_name] = old_state_name
+
+            # The case where a new state is added. When this happens, there
+            # won't be a corresponding state name in the older state dict.
+            if old_state_name not in old_states:
+                trainable_states_dict[
+                    'state_names_with_changed_answer_groups'].append(state_name)
+                continue
+            old_state = old_states[old_state_name]
+            old_training_data = old_state.get_training_data()
+            new_training_data = new_state.get_training_data()
+            if new_training_data == old_training_data and (
+                    new_state.interaction.id == old_state.interaction.id):
+                trainable_states_dict[
+                    'state_names_with_unchanged_answer_groups'].append(
+                        state_name)
+            else:
+                trainable_states_dict[
+                    'state_names_with_changed_answer_groups'].append(
+                        state_name)
+
+        trainable_states_dict['state_names_mapping'] = state_names_mapping
```

Comment:
Oh, no, don't do this please -- this is an example of scope creep. The new key/value pair has nothing to do with the point of this method, and should therefore not be returned here as part of the result dict. It should be a separate method on the exploration object and it should be tested separately. If you need to, feel free to pass in the state name mapping to the get_trainable_states_dict() method.